ably/realtime_channel.go

```diff
@@ -794,11 +808,13 @@ func (c *RealtimeChannel) notify(msg *protocolMessage) {
 		if msg.Flags != 0 {
 			c.setModes(channelModeFromFlag(msg.Flags))
 		}
-		c.Presence.onAttach(msg)
-
-		isAttachResumed := msg.Flags.Has(flagResumed)
-		if c.state != ChannelStateAttached || !isAttachResumed { //RTL12
-			c.setState(ChannelStateAttached, newErrorFromProto(msg.Error), isAttachResumed)
+		// RTL12
+		if c.state == ChannelStateAttached && !msg.Flags.Has(flagResumed) {
+			c.emitErrorUpdate(newErrorFromProto(msg.Error), false) // RTL12
+			c.Presence.onAttach(msg, false)
+		} else {
+			c.Presence.onAttach(msg, true)
+			c.setState(ChannelStateAttached, newErrorFromProto(msg.Error), msg.Flags.Has(flagResumed))
```

Comment:
This conditional looks kinda weird. You'll be in this `else` clause if either of the two conditions in the if fail, but the stuff in the else clause only makes sense if it was the first condition that failed -- eg what if you get an attached with resumed=true while you were already attached? seems like it should be
```
		if c.state == ChannelStateAttached {
			if !msg.Flags.Has(flagResumed) {
				c.emitErrorUpdate(newErrorFromProto(msg.Error), false) // RTL12
				c.Presence.onAttach(msg, false)
			}
		} else {
			c.Presence.onAttach(msg, true)
			c.setState(ChannelStateAttached, newErrorFromProto(msg.Error), msg.Flags.Has(flagResumed))
		}
```

Reply:
> I think RTL12 should mention explicitly, if attached is received on already attached channel and resume set to true, then it should do nothing

It already does say that. "If and only if the resumed flag is false, this should result in the channel emitting an UPDATE event with a ChannelStateChange object". "If _and only if_"

Reply:
```The library must never emit a ChannelState ChannelEvent for a state equal to the previous state``` means it should emit `UpdateEvent` for prevState = newState` instead of `ChannelEvent`. That's what I understood by it.

Reply:
As per RTL2f =>
```
When a channel ATTACHED ProtocolMessage is received, the ProtocolMessage may contain a RESUMED bit flag indicating that the channel has been resumed. The corresponding ChannelStateChange (either ATTACHED per RTL2a, or UPDATE per RTL12) will contain a resumed boolean attribute with value true if the bit flag RESUMED was included ...
```
I feel there's a gray area here. It doesn't mention if it's a ATTACHED on non-attached channel or ATTACHED on already attached channel. I can only assume if it's ATTACHED on already ATTACHED with resume flag as true, as per RTL2g, it should emit update event on the channel.
RTL12 only describes ATTACHED on already ATTACHED with resume flag as false, other specs are implicitly pointing towards setting a state and emitting update event for the same.

Reply:
I have updated the code according to your suggestion.
But, I would be honest, I had to read related spec multiple times again and again for this. Still, as you can see, understanding wasn't reflected in the code.
Maybe I will update the spec here https://github.com/ably/specification/pull/167 and we can resume discussion for the same.

Reply:
I have added explicit commit to the specification PR https://github.com/ably/specification/pull/167/commits/248bbc63a29e7393e56c2b72b92ab5097b628e30

Reply:
Since the change has been made as per suggestion, I am closing the issue